aot/src/main/java/com/dylibso/chicory/aot/AotMethodInliner.java

```diff
@@ -0,0 +1,82 @@
+package com.dylibso.chicory.aot;
+
+import static com.dylibso.chicory.aot.AotUtil.internalClassName;
+import static org.objectweb.asm.Type.getInternalName;
+
+import com.dylibso.chicory.wasm.exceptions.ChicoryException;
+import java.io.IOException;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.commons.ClassRemapper;
+import org.objectweb.asm.commons.Remapper;
+
+final class AotMethodInliner {
+
+    private AotMethodInliner() {}
```

Comment:
The JVM should inline the methods. The purpose of these methods is to keep the bytecode smaller and simpler, and also keep the compiler simpler. Larger bytecode prevents the JVM from optimizing. Many of the methods are there to avoid stack shuffling in the generated code.

For example, when compiling the `MEMORY_COPY` instruction, at the point when we see the instruction, we already have the `destination`, `offset`, and `size` arguments on the stack. Calling the `copy()` method on the `Memory` class requires that the `Memory` instance reference is pushed first, as the receiver (`this`) is always the first parameter to an instance method. There's no way to simply insert it that far up the stack. You have to spill all the arguments to local slots, push the receiver, then restore the arguments from locals. This would make the  bytecode far larger and require JVM optimizations to make it efficient. Instead, we simply call a static method that has the parameters in the desired order, which trivially delegates to the target method.

An alternative would be to have a far more complicated compiler that can push the receiver or other arguments at the appropriate place, but I have no idea how to do the analysis requires to make that work.

Reply:
Thanks for taking the time to do the write-up!
Your reasoning seems solid, let's move forward on this path and revisit only in presence of evidence.